BigInt support

calebsander opened the pull request:
<!--
Thank you for submitting a pull request!

Here's a checklist you might find useful.
* [ ] There is an associated issue that is labeled
  'Bug' or 'help wanted' or is in the Community milestone
* [ ] Code is up-to-date with the `master` branch
* [ ] You've successfully run `jake runtests` locally
* [ ] You've signed the CLA
* [ ] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
-->

Fixes #15096.

**Additions**:
- The `bigint` primitive type, which is analogous to but distinct from `number`
- Parsing of `bigint` literals, e.g. `123n`. Per the ECMAScript proposal, literals can be expressed in binary (`0b`), octal (`0o`), hex (`0x`), or decimal formats. `bigint` literals, like `number`, `string`, or `boolean` literals, are valid types. The config option `experimentalBigInt` must be set in order to use `bigint` literals.
- Arithmetic operators work on `bigint` values and return `bigint` results
- Updated the return type of the (JavaScript) `typeof` operator to include `"bigint"`
- `esnext.bigint` lib file, which defines the `BigInt`, `BigInt64Array`, and `BigUint64Array` types, the `BigInt` global, and the `*int64*` methods of `DataView`
- 6 tests: 
  - `parseBigInt` to test parsing `bigint` literals
  - `numberVsBigIntOperations` to test that arithmetic operations accept and return the correct types
  - `warnExperimentalBigIntLiteral` to test that using `bigint` literals without the `--experimentalBigInt` option throws an error
  - `bigintWithLib` and `bigintWithoutLib` to test the `esnext.bigint` lib file
  - `bigintIndex` to assert that `bigint` values can't be used as indices

**Notes**:

`bigint` is now a reserved word.

The type of arithmetic operations with `any` types, e.g. `<any>1 * <any>2`, was left as `number` instead of changing it to the more accurate `number | bigint` in order to maintain compatibility. Binary arithmetic operations are not allowed on operands of `number | bigint` type because it is possible that their runtime types conflict.

Error messages referring to the type of arithmetic operands were updated, as was the type of `typeof` expressions, which required updating many reference baselines. Other than that, no existing tests' expected behaviors were changed. (See commits fd5771a9c980904a4bb00edd90b62f272cde8d41 and 73d0fbf2df75b2bf8354c17127d5daa1b99a2483.)

`bigint` is not allowed as an index type. Use a `string` index instead and explicitly convert the `bigint` to a `string`.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=25886) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=25886) before we can accept your contribution.<br/><br/>:x: calebsander [sign now](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=25886)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/TypeScript?pullRequest=25886) it.</sub>

calebsander commented:
The lint is failing because `tslint` doesn't know about the `esnext.bigint` library. Seems like a chicken-and-egg situation: this push is needed to add the `esnext.bigint` library, but `tslint` needs to already have the changes in order to lint successfully.

calebsander commented:
Also not sure how to avoid running the `bigint` tests on versions of Node.js below 10, since they will fail to compile.

Kingwl commented:
cool, a big work

calebsander commented:
It is actually only a ~1000 line change, ignoring the rebuilt LKG and the updates to the reference baselines. If you would prefer it broken up, I could do the following:

- To avoid needing to rebuild the LKG in order to compile the compiler, type-alias `bigint` to `number` in the compiler code and convert all `typeof x === "bigint"` instances to `typeof x as string === "bigint"`
- Push the `esnext.bigint` lib separately, since it is not currently being used in the compiler code

Let me know whether it would be better to do this as 2 or 3 pulls instead of 1.

calebsander commented:
I reverted the LKG for now and removed references to `bigint` in the compiler code. The compiler should emit the same JS, but its typings are not as accurate. Once the LKG is rebuilt with this code, 74fd6d3 should be reverted so that the compiler code has the correct types.

mhegazy commented on src/compiler/factory.ts:
```diff
@@ -75,6 +75,11 @@ namespace ts {
         if (typeof value === "number") {
             return createNumericLiteral(value + "");
         }
+        if (typeof value as string === "bigint") {
```
we need to remove the cast, can you create a list in the issue to track changes after LKG updates.

mhegazy replied:
I do not think we should be checking the typeof the value. we should be able to compile programs with `bigint` even if running on older engines. 

I think you should revert all `typeof ... = "bigint"` changes in the code.

mhegazy commented on src/compiler/types.ts:
```diff
@@ -1,3 +1,6 @@
+// TODO: remove once tslint allows tsconfig.json to include lib "esnext.bigint"
+declare const BigInt: ((value: number | string) => number) | undefined;
```
not sure i understand why we need this in the compiler.

mhegazy replied:
I would say just use the internal representation always. either ways, this should be defined close to where it is used, and should be marked as `/* @internal */`

calebsander replied:
I may not have explained clearly how `bigint` literals are represented internally. They are stored as `{negative: boolean, base10Value: string}`, where `negative` is true iff the value is less than 0 and `base10Value` is the value of the literal as a decimal string. For example, `0xFFn` would be represented as `{negative: false, base10Value: "255"}`. I chose to store the value in decimal so that type strings) can easily be generated (e.g. `const a = 0xFFn` would declare `a` as type `255n`. It also unifies literals expressed in different bases, so `0b10000n`,`0x10n`, and `16n` are all considered the same type. This representation is used regardless of whether the `BigInt` global is defined.

`BigInt` is only used (if it is available) to convert literal strings in different bases to the decimal string. So `0xFFn` would be parsed using `"" + BigInt("0xFF")`. If `BigInt` is not available, the code falls back to computing a `Uint16Array` with the bits of the literal's value, and then computing the decimal digits of this value by repeatedly dividing by 10 and storing the remainder.

I will move the declaration to where it is used and mark it as `internal`.

mhegazy commented on src/compiler/types.ts:
```diff
@@ -3782,6 +3793,10 @@ namespace ts {
     }
 
+    export interface BigIntLiteralType extends LiteralType {
+        value: number;
```
how is it a `number`? a program using `bigint` needs to compile even if the underlying engine at compile time does not support bigint.

calebsander replied:
It was `bigint` initially, but it is `number` for now so that it can be compiled on the old LKG. See 74fd6d30806ad307cc12e432943318007838df2a.

calebsander commented on src/compiler/types.ts:
```diff
@@ -3652,60 +3659,63 @@ namespace ts {
+        ContainsObjectLiteral   = 1 << 30,  // Type is or contains object literal type
         /* @internal */
-        ContainsAnyFunctionType = 1 << 29,  // Type is or contains the anyFunctionType
+        ContainsAnyFunctionType = 1 << 31,  // Type is or contains the anyFunctionType
```
Hmm, any suggestions for how to fix that?

weswigham replied:
Usually we extract a set of flags into a new field that only exists on some subset of types. Similar to `ObjectFlags`. Given the number of them, I think literals are _probably_ ripe for this.

calebsander replied:
Presumably this could be done as a separate pull request. I'd rather avoid growing this one since it's already quite large.

sheetalkamat replied:
Can you please create issue tracking this. Thanks

calebsander replied:
#28354 is tracking this now

mhegazy commented on src/services/completions.ts:
```diff
@@ -198,7 +198,8 @@ namespace ts.Completions {
 
-    const completionNameForLiteral = JSON.stringify;
+    const completionNameForLiteral = (literal: string | number) =>
+        typeof literal as string === "bigint" ? literal + "n" : JSON.stringify(literal);
```
same here, please add this to a list of changes after LKG.

mhegazy replied:
still not sure how this is supposed to work without `bigint` support from the underlying engine?

mhegazy replied:
I think the right thing to do here is to return a richer data structure than `number | string`. since `bigint` is not that common, i would make `literals` `number | string | { bigintValue: string }`

weswigham commented on src/compiler/checker.ts:
```diff
@@ -8915,13 +8949,31 @@ namespace ts {
+                error(node, Diagnostics.Cannot_evaluate_bigint_literal_because_the_compiler_is_running_in_an_environment_without_bigint_support);
+                return +value;
+            }
+            return BigInt(value);
```
We talked about this internally when working on some other stuff; but I think similarly to our reluctance to rely on host features for any potential regex types, it's much too soon for us to rely on a host's `BigInt` constructor to be present to handle `BigInt` types in the compiler. Unlike the runtime, we don't need to perform arithmetic or much of anything on these types - the only operation we (potentially) need is equivalence. Our scanner already does octal/hex -> value conversions; we should just be able to extend that to handle integer overflow/use a psudobigint representation for large numbers and do our equivalence checking internally with a custom partial bigint.

calebsander replied:
If `bigint` literals remain valid types, you then I think you need some sort of arbitrary-precision integer library in order to compare ones in different bases, as literals can overflow the 53-bit integer available with the `number` type. It would probably be safe to just compute the lower 32 bits or something, but that isn't technically correct. Also, `evaluate` attempts to evaluate arithmetic operations on numeric literals, so do you think that should just be disabled for `bigint` literals?

calebsander replied:
Perhaps restrict `bigint` literals to the range which can be represented absolutely by the `number` type?

weswigham replied:
@calebsander You don't need a platform primitive for that, though. We read everything in as strings; we can read a bigint in as a string, then convert its value into a buffer of bytes (or another mroe efficient representation, since bigints tend to be sparse) _ourselves_, then compare those buffers, rather than relying on the host.

weswigham replied:
> Also, evaluate attempts to evaluate arithmetic operations on numeric literals, so do you think that should just be disabled for bigint literals?

Yeah, I'd just leave that off - that's just for enums, and const bigint enums aren't a pressing concern.

weswigham replied:
I'd also say that we want to be lazy here if it all possible - if a user only ever writes `0n`, then string comparisons are all we ever need - it's only when both `0x0n` and `0o0n` see use, for example, that we need to actually start parsing and comparing literal values. Necessary, for sure, but we _really_ want to avoid doing it for numbers outside the typical double range. Actually, since it seems like bigints don't support scientific notation (probably to avoid the whole negative exponent issue), there's only ever exactly 4 possible strings representing a given number - the base 2 string, the base 8 string, the base 16 string, and the base 10 string. I think if we normalize a bigint "value" internally as "the base 64 string representation of the value" (base64 simply because it's a pretty common way to encode buffers into strings - maybe another base with compression is better for memory; without profiling on a base that makes heavy use of large literals I wouldn't know), then string equality will suffice; we just need to bulk up the scanner to be able to perform that conversion.

weswigham commented:
Ah, another general remark: It looks like your parse handles it, but you need tests of bigint literals of all shapes and sizes with various configurations of numeric separators. I'd just add new entries to the numeric separator tests, myself, since those contain all kinds of literals with numeric separators in them.

sheetalkamat commented:
@mhegazy had mentioned that we should enable bigInt only under --experimentalBigInt compiler options flag.

calebsander commented:
What features would you imagine being restricted without the flag? Just `bigint` literals? The fact that `typeof` can return `"bigint"`? The `bigint` type keyword? Compiling the `esnext.bigint` lib requires the `bigint` keyword to be recognized, so if that is disabled without the flag, the `tsconfig.json` file will have to be modified in order to build the lib. I had issues with `tslint` failing on new options in `tsconfig.json`, so I'm not sure how that would work.

I'm not sure an experimental flag is necessary; no existing code should break with the change unless it is using the `bigint` identifier. I'm also not convinced "experimental" really makes sense; this doesn't seem like a feature whose behavior is likely to change.

mhegazy commented:
> What features would you imagine being restricted without the flag? Just bigint literals? 

yes. basically any expression of type `bigint` should be flagged. 
I should add that we want the error reported only once in a compilation to avoid noise.

> I'm not sure an experimental flag is necessary;

The flag is not necessary for valid code generation. the flag is needed to inform the user that there is potential change in semantics that could happen in a future release. that is the same as decorators. 
The alternative here is not to include this feature until it reaches stage 4.

kitsonk commented:
Just chipping in from the sides.  While the proposal is at Stage 3, [by TC39's definition](https://tc39.github.io/process-document/), there is a risk of implementation challenges which can effect the spec.  For example, [this behaviour](https://github.com/tc39/proposal-bigint/issues/160) isn't fully settled, while it unlikely to impact the TypeScript implementation, the flag is the only way to underline to users that it isn't fully baked yet and until it has an ES release there is a danger of how it gets aligned.  It _should_ be ES2019 but...

For something like this, it feels fairly important to flag it until the proposal reaches Stage 4.

calebsander commented:
Sure, that makes sense. I already added the flag. It will warn on any `bigint` literals, or if it sees the `bigint` keyword. I think I've addressed all the comments on the PR, with the exception of `TypeFlags` exceeding 30 bits. I looked into trying to fix it by factoring out the literal flags, but it looks like there are hundreds of locations that would have to be updated. I would rather keep that for a separate PR. I think this PR should be ready to merge if there are no further comments.

The Node.js 8 Travis job seems to have failed on the last build because npm couldn't resolve the `jake` executable. Not sure what would cause that. Otherwise, the builds are all passing.

weswigham commented:
> I already added the flag. It will warn on any bigint literals, or if it sees the bigint keyword. 

I think we don't want to error in the keyword, just the literals. We don't want to error when the node .d.ts updates to use bigints in stats, for example - just on usages we won't be downleveling.

mhegazy commented:
> I think we don't want to error in the keyword, just the literals. We don't want to error when the node .d.ts updates to use bigints in stats, for example - just on usages we won't be downleveling.

That sounds right to me. in general we should not error in an ambient context any ways..

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -80,6 +80,9 @@ namespace ts {
         let apparentArgumentCount: number | undefined;
 
+        /** Whether user has not yet been warned about using BigInt without --experimentalBigInt */
+        let notWarnedAboutBigInt = true;
```
OK, so after talking this one out with other ppl on the team, I believe this was a stupid suggestion on my side. 

IDEs ask for errors on files in no specific order. with this in place the error will keep appearing and disappearing randomly. i think we are better off having the error on every use. We already report errors on every use for decorators. so this should just be the same. I think we can just remove this check, and keep the error.

Sorry for the bad suggestion.

calebsander commented:
Hmm, it seems like the Travis build failure on Node 8 is due to `jake` not being installed in the `node_modules` cache, since it went away when I removed the `cache` field from `.travis.yml`. Not sure what can be done to fix that, but I don't think there is any problem with the PR code.

sheetalkamat commented:
Also need testcase for usage of bigint literals, type and BigInt() when esNext and when less than esNext with and without experimentalBigInt

DanielRosenwasser commented on src/compiler/binder.ts:
```diff
@@ -3631,9 +3631,13 @@ namespace ts {
+                if (flags & TokenFlags.BinaryOrOctalSpecifier) {
                     transformFlags |= TransformFlags.AssertES2015;
                 }
+                if (flags & TokenFlags.BigInt) {
```
I feel like instead of putting this in `NumericLiteral`, you should have a new `BigIntLiteral` type; not sure if others on the team agree with me here.

calebsander replied:
It was easier to parse them both as `SyntaxKind.NumericLiteral` expressions since many of the same flags still apply (namely `HexSpecifier`, `BinarySpecifier`, `OctalSpecifier`, `ContainsSeparator`). I split them once they were converted into `LiteralType`s.

sheetalkamat replied:
I think i agree with @DanielRosenwasser .. Though the flags are same, you could create the SyntaxKind.BigIntLiteral making it easier to ensure they arent treated like numbers unknowingly. (note the type of number and bigint is different and hence explicit token kind is justified)

calebsander commented:
> Also need testcase for usage of bigint literals, type and BigInt() when esNext and when less than esNext with and without experimentalBigInt

`experimentalBigInt` and `"target": "esnext"` vs. pre-`esnext` are currently orthogonal options: `experimentalBigInt` controls whether errors are raised on `bigint` literals, whereas `target` only affects whether the `esnext.bigint` lib is included. Regardless of both options, `bigint` can be used as a type without raising any errors. (I'm not sure whether it makes sense to allow the `bigint` keyword without `experimentalBigInt`, but I think it should behave the same in ambient and non-ambient settings. I think @weswigham's point about not throwing errors just because `@types/node` is included is really more about included libraries vs. user-written code than about ambient vs. non-ambient contexts.)

The `parseBigInt` and `warnExperimentalBigIntLiteral` tests already verify the behavior of the having `experimentalBigInt` set and unset. I added tests verifying the behavior of variables and types declared in `esnext.bigint` with and without `"target": "esnext"`, including the non-primitive resolution of `bigint` (to `BigInt` with `"target": "esnext"`, and to `{}` without `esnext`).

sheetalkamat commented on src/compiler/checker.ts:
```diff
@@ -7135,6 +7161,7 @@ namespace ts {
             return t.flags & TypeFlags.Intersection ? getApparentTypeOfIntersectionType(<IntersectionType>t) :
                 t.flags & TypeFlags.StringLike ? globalStringType :
                 t.flags & TypeFlags.NumberLike ? globalNumberType :
+                t.flags & TypeFlags.BigIntLike ? getGlobalBigIntType(/*reportErrors*/ languageVersion >= ScriptTarget.ESNext) :
```
sorry i didnt see the `||emptyObjectType` earlier..

weswigham commented:
So, question raised in #26330 - should `bigint`s be allowed to index objects/arrays (probably?) and in that case, shouldn't they just be treated as `number`s (since they have the same string representation)? And should such an addition affect the base `keyof` (eg, should `keyof`'s constraint be `string | symbol | number | bigint`)?

calebsander commented:
Hmm, seems like kind of a slippery slope. Any value can technically be used as a key in JS, since any value can be converted to a string representation:
````javascript
const a = {}
a[new Date] = 'abc'
a[function abc() {}] = 'abc'
a[[1, 2, 3]] = 'abc'
console.log(a)
/* Prints:
{ 'Fri Aug 10 2018 11:50:27 GMT-0700 (Pacific Daylight Time)': 'abc',
  'function abc() {}': 'abc',
  '1,2,3': 'abc' }
*/
````

I think key types only include `number` because there are several builtins (`Array`, `TypedArray`s, etc.) which are meant to be indexed with numeric indices. All of these can technically be indexed with `bigint` indices as well, but that is only because the `bigint` gets converted to its string representation first. This There are no instances in the standard library of types expressly indexed by `bigint`. I can't think of any situation in which a user-defined type would want to have a `bigint` index either. You could also always just declare it as a `string` index and then convert the `bigint` to a `string` before using it as an index. So I would probably lean towards not including `bigint` as an index type.

sheetalkamat commented:
Agreed with @calebsander on https://github.com/Microsoft/TypeScript/pull/25886#issuecomment-412175959 where in we shouldn't include bigint in number signature. BigInt shouldn't affect signature (or keyof at all).

weswigham commented:
Alright, I'm fine with that - just wanted to bring it up. Also should probably add tests to document that we do error on such uses.

DanielRosenwasser commented:
Just as a tip @calebsander, some of your commits aren't associated with your account. You may want to update your email addresses with GitHub.

weswigham commented on src/compiler/utilities.ts:
```diff
@@ -8253,4 +8255,60 @@ namespace ts {
+
+        let log2Base: number;
+        if (stringValue.startsWith("0b")) log2Base = 1;
+        else if (stringValue.startsWith("0o")) log2Base = 3;
```
These base indicator comparisons need to be case-insensitive. `0O` and `0X` are afaik all still allowed on bigints just like normal ints. personally, I think slicing the first two characters and switching on them to determine the base is a bit nicer looking, but that's a matter of preference.

calebsander replied:
`0b`, `0o`, and `0x` are not parsed directly from the code; the scanner always lower-cases the base prefix. See https://github.com/Microsoft/TypeScript/pull/25886/files/2edab35e278984ae7f30ff28a2698a64a7e390d5#diff-a15280067dcb3b0929808762f8352e86R1572. I'll add tests for it though.

weswigham replied:
Right, but there's nothing enforcing that invariant in the scanner and it's reasonable to expect that someone calls this with the raw text of a literal at some point.

weswigham commented on src/compiler/utilities.ts:
```diff
@@ -8253,4 +8255,60 @@ namespace ts {
+
+    // Converts a bigint literal, e.g. `0x1234n`,
+    // to its decimal representation, e.g. `4660`.
+    export function parsePseudoBigInt(stringValue: string): string {
```
Just for prudence, I feel like this is the kind of method that'd be nice to be unittested on its own; but it's probably not strictly required since other tests should exercise it enough.

calebsander replied:
Added a unit test: https://github.com/Microsoft/TypeScript/pull/25886/files#diff-b222e6a526e22d4babe56a51168f5fae

weswigham commented:
@sheetalkamat you wanna help guide this to a merge for the next release?

weswigham commented:
@calebsander care to sync with `master`?

calebsander commented:
Synced.

sheetalkamat commented on src/compiler/checker.ts:
```diff
@@ -555,61 +559,71 @@ namespace ts {
             FunctionFacts = FunctionStrictFacts | EQUndefined | EQNull | EQUndefinedOrNull | Falsy,
-            UndefinedFacts = TypeofNEString | TypeofNENumber | TypeofNEBoolean | TypeofNESymbol | TypeofNEObject | TypeofNEFunction | TypeofNEHostObject | EQUndefined | EQUndefinedOrNull | NENull | Falsy,
+            UndefinedFacts = TypeofNEString | TypeofNENumber | TypeofNEBigInt | TypeofNEBoolean | TypeofNESymbol | TypeofNEObject | TypeofNEFunction | TypeofNEHostObject | EQUndefined | EQUndefinedOrNull | NENull | Falsy,
             NullFacts = TypeofEQObject | TypeofNEString | TypeofNENumber | TypeofNEBoolean | TypeofNESymbol | TypeofNEFunction | TypeofNEHostObject | EQNull | EQUndefinedOrNull | NEUndefined | Falsy,
```
Should contain `TypeofNEBigInt` ?

sheetalkamat commented on src/compiler/parser.ts:
```diff
@@ -2902,6 +2901,7 @@ namespace ts {
                 case SyntaxKind.UnknownKeyword:
                 case SyntaxKind.StringKeyword:
                 case SyntaxKind.NumberKeyword:
+                case SyntaxKind.BigIntKeyword:
```
In parseJsonText you need to handle numeric literal checks (to ignore the big ints)

calebsander replied:
I separated `SyntaxKind.BigIntLiteral` from `SyntaxKind.NumericLiteral`, so I believe that is no longer necessary

sheetalkamat commented on src/compiler/parser.ts:
```diff
@@ -2960,6 +2960,7 @@ namespace ts {
                 case SyntaxKind.UnknownKeyword:
                 case SyntaxKind.StringKeyword:
                 case SyntaxKind.NumberKeyword:
+                case SyntaxKind.BigIntKeyword:
```
Need to handle bigint in `parsePropertyNameWorker`

calebsander replied:
I don't think `bigint`s are valid property names, e.g. `{1n: 2}` throws a `SyntaxError` in V8

sheetalkamat commented on tests/cases/compiler/bigintIndex.ts:
```diff
@@ -0,0 +1,25 @@
+// Show correct usage of bigint index: explicitly convert to string
+const bigNum: bigint = 0n;
+const typedArray = new Uint8Array(3);
+typedArray[bigNum] = 0xAA; // should error
```
Need another test for class property and methods with bigint as name (shouldn't be allowed)

sheetalkamat commented on src/compiler/commandLineParser.ts:
```diff
@@ -583,6 +584,12 @@ namespace ts {
+            name: "experimentalBigInt",
+            type: "boolean",
+            category: Diagnostics.Experimental_Options,
+            description: Diagnostics.Enables_experimental_support_for_ESNext_BigInt_expressions
```
Handle bigint as key in ParsedJson to report error: `convertPropertyValueToJson` add test as well.

sheetalkamat replied:
Handle type `JsonMinusNumericLiteral`

sheetalkamat commented on src/compiler/binder.ts:
```diff
@@ -3697,9 +3697,13 @@ namespace ts {
+                if (flags & TokenFlags.BigInt) {
+                    transformFlags |= TransformFlags.AssertESNext;
+                }
                 break;
```
Do we need to handle `checkStrictModeNumericLiteral` for bigInt?

calebsander replied:
Don't think there are any differences between `bigint` literals in strict and non-strict mode.

sheetalkamat commented on tests/cases/compiler/bigintIndex.ts:
```diff
@@ -0,0 +1,25 @@
+key = Symbol();
+key = 123n; // should error
+
+// Show correct usage of bigint index: explicitly convert to string
```
Are bigints allowed as object literal binding element property as is? Test for the same

sheetalkamat replied:
`getPropertyNameOfBindingOrAssignmentElement` need to be handled with bigint. Noting this for tracking purpose

sheetalkamat commented on src/compiler/checker.ts:
```diff
@@ -21292,7 +21367,7 @@ namespace ts {
 
         function checkArithmeticOperandType(operand: Node, type: Type, diagnostic: DiagnosticMessage): boolean {
-            if (!isTypeAssignableToKind(type, TypeFlags.NumberLike)) {
+            if (!isTypeAssignableTo(type, numberOrBigIntType)) {
```
I think you can change this so as to optimize? : 
```ts
if(isTypeAssignableToKind(type, TypeFlags.NumberLike)) {
return true;
}
if(isTypeAssignableTo(type, bigIntType)) {
return true;
}
// error here
```

sheetalkamat commented on src/compiler/checker.ts:
```diff
@@ -21440,10 +21515,19 @@ namespace ts {
+                            ? -value
+                            : { negative: !value.negative, base10Value: value.base10Value }
+                    ));
                 }
```
In `checkGrammarNumericLiteral` you need to error if `options.experimentalBigInt` is not true and language is not `esnext` Note this is better place to error this than `getNumericLiteralValue`

sheetalkamat commented on src/compiler/checker.ts:
```diff
@@ -21440,10 +21515,19 @@ namespace ts {
-                    return getFreshTypeOfLiteralType(getLiteralType(+(<LiteralExpression>node.operand).text));
+                    if (isTypeAssignableToKind(operandType, TypeFlags.BigIntLike)) {
+                        error(node.operand, Diagnostics.Operator_0_cannot_be_applied_to_type_1, tokenToString(node.operator), typeToString(operandType));
+                    }
```
Need to handle evaluation of bigInt as enumMember value and a test. `computeConstantValue` function

sheetalkamat commented on src/compiler/checker.ts:
```diff
@@ -26473,7 +26605,8 @@ namespace ts {
                         checkGrammarNumericLiteral(<NumericLiteral>expr);
-                        return +(<NumericLiteral>expr).text;
+                        const literalValue = getNumericLiteralValue(<NumericLiteral>expr);
+                        return typeof literalValue === "number" ? literalValue : undefined; // don't evaluate bigint
```
Need to handle evaluation of bigInt as enumMember value and a test. Cannot return undefined.. Need to add test for `const enums` with member initializers and operators.

ajafff commented on src/harness/fourslash.ts:
```diff
@@ -4838,6 +4839,7 @@ namespace FourSlashInterface {
             "unknown",
             "from",
             "global",
+            "bigint",
```
it actually contains type keywords like `unknown`, `never` and so on

sheetalkamat commented on src/compiler/transformers/ts.ts:
```diff
@@ -2006,6 +2006,9 @@ namespace ts {
                 case TypeReferenceSerializationKind.VoidNullableOrNeverType:
                     return createVoidZero();
 
+                case TypeReferenceSerializationKind.BigIntLikeType:
```
You also need to handle SyntaxKind.BigIntKeyword and BigIntLiterals in `serializeTypeNode`

sheetalkamat commented on tests/cases/compiler/bigintIndex.ts:
```diff
@@ -0,0 +1,25 @@
+const typedArray = new Uint8Array(3);
+typedArray[bigNum] = 0xAA; // should error
+typedArray[String(bigNum)] = 0xAA;
+typedArray["1"] = 0xBB;
```
Need test cases when bigInt is used as property name and type has index signature: 1. string, 2.number 3. both 4. no signature

sheetalkamat replied:
Also need tests with these as contextual type

sheetalkamat commented on src/compiler/transformers/ts.ts:
```diff
@@ -2006,6 +2016,9 @@ namespace ts {
                     return createVoidZero();
 
+                case TypeReferenceSerializationKind.BigIntLikeType:
+                    return createIdentifier("BigInt");
```
This needs to be 
```ts
return languageVersion < ScriptTarget.ESNext 
                        ? getGlobalBigIntNameWithFallback() 
                        : createIdentifier("BigInt");
```

sheetalkamat commented on src/compiler/utilities.ts:
```diff
@@ -8433,4 +8446,88 @@ namespace ts {
+     */
+    export function parsePseudoBigInt(stringValue: string): string {
+        // Use native BigInt if available
+        if (typeof BigInt !== "undefined") {
```
Can you please get rid of this. We would like to be unifying and always depend on our own string that will give consistent results in all cases.

sheetalkamat commented on src/compiler/checker.ts:
```diff
@@ -30336,6 +30476,13 @@ namespace ts {
             return false;
         }
```
In `checkAmbientInitializer` you need to `isInvalidInitializer` also need to be false if `node.initializer` is `BigIntLiteral`

sheetalkamat replied:
And `PrefixUnary` with bigint literal. (note you dont want to change `isStringOrNumberLiteralExpression` since its used for enum member access as well but add one similar to that)

calebsander replied:
I believe `isStringOrNumberLiteralExpression(node.initializer) || isSimpleLiteralEnumReference(node.initializer) || node.initializer.kind === SyntaxKind.TrueKeyword || node.initializer.kind === SyntaxKind.FalseKeyword` will already be false if `node.initializer.kind === SyntaxKind.BigIntLiteral`. Am I wrong?

sheetalkamat replied:
I meant it should be true if node.initializer is bigIntLiteralExpression. (we want to retain `const x = 12n` in dts just like `const x = 12`) Add that as a test as well.

sheetalkamat commented on src/compiler/checker.ts:
```diff
@@ -9256,13 +9292,24 @@ namespace ts {
                     const name = prop.valueDeclaration && getNameOfDeclaration(prop.valueDeclaration);
-                    type = name && isNumericLiteral(name) ? getLiteralType(+name.text) :
-                        name && name.kind === SyntaxKind.ComputedPropertyName && isNumericLiteral(name.expression) ? getLiteralType(+name.expression.text) :
+                    type = name && isNumericLiteral(name) ? getNumericLiteralType(name) :
```
Can you please revert this. And inline all numericLiteralType creation as was earlier. Thanks.

sheetalkamat commented on tests/baselines/reference/parseBigInt.errors.txt:
```diff
@@ -0,0 +1,124 @@
+==== tests/cases/compiler/parseBigInt.ts (17 errors) ====
+    // All bases should allow "n" suffix
+    const bin = 0b101, binBig = 0b101n; // 5, 5n
+    const oct = 0o567, octBig = 0o567n; // 375, 375n
```
octals are error right?

calebsander replied:
Only legacy octals, e.g. `0567n`. `0o567n` is allows.

sheetalkamat replied:
But then I don't see test case for that. Please add one.

calebsander replied:
Already have one here: https://github.com/Microsoft/TypeScript/pull/25886/files/0d54f041d66536b5712e9bb5e94f28aa7d430892#diff-f6f1b31177fd1675c39b4b3f7ad83581R80

sheetalkamat replied:
The pointer is wrong. can you please provide correct link Thanks. Unresolving this comment so that i can track this.

calebsander replied:
Ah, sorry about that. Must be because I force-pushed. I'll copy the section here, from the `parseBigInt` test:
````typescript
{ const legacyOct = 0123n; }
````

sheetalkamat commented on tests/cases/compiler/warnExperimentalBigIntLiteral.ts:
```diff
@@ -0,0 +1,6 @@
+
+const normalNumber = 123; // should not error
+let bigintType: bigint; // should not error
+let bigintLiteralType: 123n; // should error when used as type
```
when using big int literal as type (eg 123n in `let bigintLiteralType: 123n` ) it should error. Please fix that

calebsander replied:
Do you mean that it should *not* error? If so, what is the recommended way of checking whether a literal is used as a type or a value? Can you use something like `isTypeNode()`?

weswigham replied:
You can check the `.parent`. If it's a literal type node, don't error. If it's a prefix unary expression, you can check that node's parent.

sheetalkamat commented on tests/cases/compiler/bigintWithLib.ts:
```diff
@@ -0,0 +1,50 @@
+// @experimentalBigInt: true
```
Can you add `@declaration: true to this test case so we can verify generation of declaration file as well.`
While doing that you want to also add 
```ts
const x = 12n; // should emit type 12n
const y : 12n = 12n; // should emit type 12n
let z = 12n; // should emit type bigint in declaration file
```

sheetalkamat commented:
Thank you @calebsander for this contribution. It should be available in tonight's nightly build.

joshuajbouw commented:
Hey @calebsander, I just updated to the Nightly and now I am having TSLint issues, but it is compiling correctly. What were the flags to remove the warnings again? I think I noticed that you deleted them, or removed them.

calebsander commented:
I'm not sure what is required for TSLint to work, but for `tsc` to succeed you should only have to set `"target": "esnext"`. This will prevent warnings for `BigInt` literals and include the `esnext.bigint` standard library declarations. The developers removed the `experimentalBigInt` flag; it's now indicated by setting the target version.

joshuajbouw commented:
"esnext.bigint", where would I find these?

Thanks.

weswigham commented:
> I do have my target already set at esnext, still not fixed on Atom or VS Code. 

You need to set vscode to use your workspace version of typescript and install the `@next` version in your workspace - this hasn't shipped bundled with the editor yet, and won't until after the 3.2 release in the coming few weeks.

calebsander commented:
> "esnext.bigint", where would I find these?

This is just the declaration file that specifies `BigInt`, `BigInt64Array`, etc. If you click `Go to Definition` in VS Code on the `BigInt` constructor, it should take you to the `esnext.bigint` file where it is defined (which should be `node_modules/typescript/lib/lib.esnext.bigint.d.ts`). As @weswigham said, you have to make VS Code use your workspace version of TypeScript. This `esnext.bigint` declaration file is automatically included when you target `esnext`, but you can also include it manually by adding it to `lib` in your `tsconfig.json`.

joshuajbouw commented:
Saviours, thank you. I hope your information will help the next person.
